gilbarbara/react-joyride
# Handle unmounted steps

## Description

This PR builds off the work done by @goloveychuk in https://github.com/gilbarbara/react-joyride/pull/167.  

This PR changes the behavior of `react-joyride` when it tries to render a tooltip or beacon for a step's selector which it cannot find in the DOM.  Rather than simply incrementing/decrementing the index to skip the step, a callback is now triggered and the joyride is stopped.  This way, it allows the application to decide what to do when a target element is not found.

An app can still accomplish the same behavior by listening to the callback and incrementing/decrementing the `stepIndex` prop as appropriate.  The demo has been updated to demonstrate exactly this.

The app could also stop the joyride and then retry after a period of time, in case it is waiting for an element to be added to its DOM.

## To Test

Click through the demo as usual, and verify that apart from seeing a console warning about a missing target, everything works as normal.

## Reviewers Should Focus On

I'm slightly nervous about the change to `componentWillReceiveProps`, because the added `if` statement is not comparing current to next props, but rather next props to current state.  I think this is good to do in this case, because if an app keeps the `run` prop turned on, and the Joyride itself stops running because it could not find a target, it's nice for the app to be able to just change the stepIndex and allow the joyride to keep going.  Without that change to `componentWillReceiveProps`, the application would have to use two render cycles -- the first to turn `run` to false, and a second to change the index and turn `run` back to true.

I think it is safe, because if the app is changing the `stepIndex` and `run` is true, I think the app should have a reasonable expectation that the joyride will run at the specified `stepIndex`.  I just wanted to call this out for review.
